src/number_theory/bernoulli.lean

```diff
@@ -258,3 +258,99 @@ begin
+  end,
+  sorry,
+end
+
+section faulhaber
+
+lemma aux_cauchy_prod (n : ℕ) :
+mk (λ (p : ℕ), bernoulli p / ↑p!) * mk (λ (q : ℕ), (coeff ℚ (q + 1)) (exp ℚ ^ n))
+ = mk (λ (p : ℕ), (range (p + 1)).sum (λ (i : ℕ),
+ bernoulli i * ↑((p + 1).choose i) * ↑n ^ (p + 1 - i) / ↑(p + 1)!)) :=
+begin
+  let f := λ a, λ b, bernoulli a / ↑a! * (coeff ℚ (b + 1)) (exp ℚ ^ n),
+  ext q,
+  simp only [power_series.coeff_mul, coeff_mk, cast_mul, nat.sum_antidiagonal_eq_sum_range_succ f],
+  refine sum_congr rfl _,
+  simp_intros m h only [finset.mem_range],
+  have hq : (q.succ! : ℚ) ≠ 0 := by exact_mod_cast q.succ.factorial_ne_zero,
+  have hqm1 : q - m + 1 = q + 1 - m,
+  { rw nat.sub_add_comm,
+    exact le_of_lt_succ h },
+  simp only [f, exp_pow_eq_rescale_exp, rescale, one_div, coeff_mk, ring_hom.coe_mk, coeff_exp,
+    ring_hom.id_apply, cast_mul, rat.algebra_map_rat_rat],
+  rw [choose_eq_factorial_div_factorial h.le, eq_comm, div_eq_iff hq, succ_eq_add_one,
+    mul_assoc _ _ ↑(q.succ!), mul_comm _ ↑(q.succ!), ←mul_assoc, div_mul_eq_mul_div,
+    mul_comm (n ^ (q - m + 1) : ℚ), ←mul_assoc _ _ (n ^ (q - m + 1) : ℚ), ←one_div, mul_one_div,
+    div_div_eq_div_mul, hqm1, cast_dvd, cast_mul],
+  { ring },
+  { exact factorial_mul_factorial_dvd_factorial h.le },
+  { simp [not_false_iff, factorial_ne_zero] },
+end
+
+lemma aux_power_series_equal (n : ℕ) :
+  (exp ℚ - 1) * power_series.mk (λ p, (range n).sum (λ k, (k : ℚ)^p * algebra_map ℚ ℚ p!⁻¹)) =
+  (exp ℚ - 1) * power_series.mk (λ p, (range (p + 1)).sum(λ i,
+  (bernoulli i) * ((p + 1).choose i) * n^(p + 1 - i)/((p + 1)!))) :=
+begin
+  have h_geom_sum : (geom_series (exp ℚ) n) * (exp ℚ - 1) = (exp ℚ)^n - 1 := geom_sum_mul _ _,
+  have h_const : C ℚ (constant_coeff ℚ ((exp ℚ)^n)) = 1,
+    { simp only [one_pow, constant_coeff_exp, ring_hom.map_pow, ring_hom.map_one], },
+  have h_r : (exp ℚ)^n - 1 = X * mk (λ (p : ℕ), (coeff ℚ (p + 1)) (exp ℚ ^ n)),
+    { rw [←h_const], rw [sub_const_eq_X_mul_shift], },
+  simp [geom_series] at h_geom_sum,
+  rw [←exp_pow_sum n, mul_comm, h_geom_sum, h_r, ←bernoulli_power_series, mul_assoc,
+    aux_cauchy_prod],
+end
+
+/-- Faulhabers' theorem: sum of powers. -/
+theorem faulhaber (n p : ℕ) :
+(range n).sum(λk, (k : ℚ)^p) =
+(range (p + 1)).sum (λ i, (bernoulli i) * ((p + 1).choose i) * n^(p + 1 - i) / (p + 1)) :=
+begin
+  have hpseq : power_series.mk (λ p, (range n).sum (λ k,
+    (k : ℚ)^p * algebra_map ℚ ℚ p!⁻¹)) =
+    power_series.mk (λ p, (range (p + 1)).sum(λ i,
+      (bernoulli i) * ((p + 1).choose i) * n^(p + 1 - i)/((p + 1)!))),
+  { have hexp : exp ℚ - 1 ≠ 0,
+    { rw [exp],
+      simp only [power_series.ext_iff, linear_map.map_zero, one_div, coeff_mk, coeff_one,
+      ring_hom.id_apply, linear_map.map_sub, ne.def, not_forall,
+      rat.algebra_map_rat_rat],
+      use 1,
+      simp only [factorial_one, sub_zero, if_false, inv_one, not_false_iff, one_ne_zero,
+        cast_one], },
+    apply mul_left_cancel' hexp (aux_power_series_equal n), },
```

Comment:
@mo271 I think it would be significantly less work if you moved this into `aux_power_series_equal` and then altered the statement of that lemma to something more desirable, so that your `hpseq` will truly be `aux_power_series_equal`. I'm not really sure how to describe what I'm saying, but if you'd like I can push a commit with the changes I am envisioning, which you can then edit as you please.

Reply:
Sure, please push!

Reply:
I pushed this commit. Here are a few points regarding additional changes I made that followed from implementing what I described above :
- In `aux_power_series_equal`, `h_const` is only used in proving `h_r`, so I tucked it inside that proof.
- Altering the statement of `aux_power_series_equal` allows me to close its proof with `simpa [...] using geom_sum_mul (exp ℚ) n`, eliminating 
```
have h_geom_sum : (geom_series (exp ℚ) n) * (exp ℚ - 1) = (exp ℚ)^n - 1 := geom_sum_mul _ _,
...
simp [geom_series] at h_geom_sum,
```
- I golfed both proofs by making style corrections, combining `simp`s/`rw`s, removing unnecessary arguments, and deleting extra parentheses. My reorganization also made some of the subproofs easier to prove, hence shorter. 
- After moving `hp` earlier in the proof, I realized that the proofs of `hp`, `hne_zero'`,  and `hfrac`, along with the `ring, rw [hfrac],` at the end of your proof, could all be accomplished by a single, terminal `field_simp`, so I made that change, too.
From the fact that `faulhaber` still contains `simp only [...] at hpseq` and `rw [...] at hpseq` (and from the extra work you are doing with `←sum_mul ...` followed by `sum_mul`) it is clear that what I have done here could be furthered, but I don't think it pays as of now because it may very well become moot depending on the final structure of your proof.

Reply:
The build fails due to something unrelated, I'm not sure why.

Reply:
cool, thanks!

Reply:
I pushed a commit that fixes the error it was snagging on, it now fails only due to the `sorry`.